Dispatch App : Change parameter semantics

andrewkaufman opened the pull request:
It seems reasonable to use the `DispatchDialogue` with no visible task nodes, in the case of a pre-made script. This allows the user to differentiate between tasks to dispatch and nodes to display.

- Added `-tasks` for specifying the task nodes to dispatch.
- Changed `-nodes` to specify which node UIs should be shown in `-gui` mode.

I've provided backwards compatibility since we have a few instances of the `gaffer dispatch` commandline in production, but I think its contained enough at IE that I could drop those commits if we'd rather just move on. I will probably need to backport this to 0.51 though.

andrewkaufman commented:
Note that we'd originally discussed this idea as an `-extraNodes` parameter to be appended to the main `-nodes` in the gui. But upon implementing that and showing it to @murrays-ie, we decided it'd be better to have 2 separates lists.

It also helps prepare us for a distant future when `Dispatchers` work with `TaskPlugs` directly rather than nodes, and we may want to dispatch one-of-several tasks on a given node.

johnhaddon commented on apps/dispatch/dispatch-1.py:
```diff
@@ -145,8 +145,12 @@ def __init__( self ) :
-			return 1
+			# fallback to nodes for backwards compatibility
+			if len( args["nodes"] ) :
+				IECore.msg( IECore.Msg.Level.Error, "gaffer dispatch", "Use the \"tasks\" parameter to specify the dispatchable nodes. The \"nodes\" parameter is only for gui purposes." )
```
Warning rather than Error?

andrewkaufman replied:
Addressed in 565b5a60f05aedeed008ec1d77411d29c03fed96

johnhaddon commented on python/GafferDispatchUI/DispatchDialogue.py:
```diff
@@ -58,15 +58,15 @@ class DispatchDialogue( GafferUI.Dialogue ) :
 	__dispatchDialogueMenuDefinition = None
 
-	def __init__( self, tasks, dispatchers, nodes, postDispatchBehaviour=PostDispatchBehaviour.Confirm, title="Dispatch Tasks", sizeMode=GafferUI.Window.SizeMode.Manual, **kw ) :
+	def __init__( self, tasks, dispatchers, nodes=None, postDispatchBehaviour=PostDispatchBehaviour.Confirm, title="Dispatch Tasks", sizeMode=GafferUI.Window.SizeMode.Manual, **kw ) :
```
This doesn't provide full backwards compatibility, because we've still added a new positional argument before `postDispatchBehaviour`. For full compatibility, the nodes argument would need to be added after all the others, which is a bit of a pity, because it's definitely a more important argument than `postDispatchBehaviour`. If you're not in desperate need of a backport, I'd favour the clean break.

As an aside, it seems that Python 3 has [keyword-only arguments](https://www.python.org/dev/peps/pep-3102/), which could help with this sort of problem. I could see myself using them not just to help with this type of problem, but also to force call sites to be more readable - it's often easier to understand keyword arguments than positional ones.

johnhaddon replied:
OK, but we can't backport without it being fully API-compatible. For master I think what we have already is ideal, so unless a backport is absolutely essential, I'd go with that..

andrewkaufman replied:
Hmm, just coming out of my vacation brain, perhaps this is non-backportable but still an improvement. In the complex case that motivated the PR, we want to specify different tasks vs UI nodes. But in the simpler case of showing the node you’re dispatching, you’d need to double-specify it without this (and the previous) commit. Thoughts?

johnhaddon replied:
I thought this was an improvement too, but thinking more about it I don't think it's quite right. There should definitely be a way of preventing any nodes from being shown, but currently there isn't, either in the DispatchDialogue or the app.

For the DispatchDialogue we could have `nodes = None` be the default, and mean "show the tasks", with the ability to pass `nodes = []` to say "show nothing". For the app it's harder, since we don't have any ability to distinguish between "no list" and an empty list. Maybe a special default of "<tasks>" could mean show the tasks, and then an empty list can be used for "show nothing"?

johnhaddon replied:
Alternatively, I think I'm actually fine with having to pass the same list to each argument if I want to show all the tasks - at least it's nice and explicit. If we did that though, we'd need to allow a `-nodes` argument even when there's no script in the dispatch app, so one can do `-tasks GafferDispatch.SystemCommand -nodes SystemCommand`.

andrewkaufman replied:
Addressed in 4aa3d189a67066e9dcf436b01c1ca4ec0ac53b90

andrewkaufman commented:
I rebased on master, dropping the problematic dialogue compatibility commit, and adding 2 new commits addressing the review comments.

I did keep the app parameter compatibility commit though, since I think it'll be useful for folks who've already writen dispach app commandlines onto SystemCommands nodes in their scripts.

johnhaddon commented on apps/dispatch/dispatch-1.py:
```diff
@@ -167,28 +167,15 @@ def _run( self, args ) :
-			node = script.descendant( nodeName )
-			if node is None :
-				IECore.msg( IECore.Msg.Level.Error, "gaffer dispatch", "\"%s\" does not exist." % nodeName )
+			node = self.__acquireNode( nodeName, script, args )
```
I don't think calling `__acquireNode()` here makes sense - we only want to show existing nodes, not create new ones (that won't be dispatched anyway). It also means that typos are reported as a confusing "Please provide the namespaced python class for..." rather than "node does not exist"...

andrewkaufman replied:
Addressed in da8b7d2603bbe8ee11719231da6db91c3e92779c

andrewkaufman commented:
Do you want me to keep -tasks and -nodes (with the later deprecated) or ditch - tasks entirely?

johnhaddon commented:
Up for discussion, but as a starting point I'd propose we ditch `-tasks` completely. I think in the long run you're right to be wanting to rename it, but I think that should be accompanied by it accepting the names of task _plugs_ rather than nodes. We have an [equivalent todo](https://github.com/GafferHQ/gaffer/blob/master/include/GafferDispatch/Dispatcher.h#L135-L139) for `Dispatcher::dispatch()`, and we'd want to also switch the `execute` app to match.

andrewkaufman commented:
Yeah the rename was an attempt to prepare for that future, changing the cli now while the app is relatively new. I’d hope in that future the -tasks arg would accept a plug or node (with the later dispatching all out tasks).

andrewkaufman commented:
> the temporary execute/dispatch mismatch isn't too important?

I'm inclined to agree with that (though I hadn't even considered it initially). It is a bit odd that they don't match, but I don't expect people to write their own `execute` commands anymore, and they are different apps, so I don't think we're obligated for them to match in the first place.

So that leaves:

- [ ] Keep `-tasks`, drop `-nodes` entirely, add `-show`
- [ ] Add a doc string demonstrating `-show`
- [ ] Fix use of `__acquireNode`

Did I miss anything?
